### PR TITLE
Boundary averages

### PR DESCRIPTION
**Pull request by ian-r-rose**
Postprocessors for calculating laterally-averaged properties at the top and bottom of the domain.

### CODE REVIEW

File: doc/modules/changes.h [hunk 1]
```diff
@@ -6,6 +6,12 @@
  *
  *
  * <ol>
+ * 
+ * <li> New: There are now postprocessors BoundaryDensities and
+ * BoundaryPressures which calculate laterally averaged densities 
+ * and pressures at the top and bottom of the domain. The intention
+ * is that they may be used for dynamic topography calculations.
+ * (Ian Rose, 2015/05/27)
```
Comment (bangerth) [inline doc/modules/changes.h]: Drop the last sentence. People come up with all sorts of reasons why they want to use stuff, no need to limit their imagination :-)

File: include/aspect/postprocess/boundary_densities.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+         */
+        double density_at_top() const;
+
+        /**
+         * Query the density at the bottom surface
+         */
+        double density_at_bottom() const;
+
+      private:
+        double top_density;
```
Comment (bangerth) [inline include/aspect/postprocess/boundary_densities.h]: add a comment that says that these variables are filled whenever execute is executed.

File: doc/modules/changes.h (hunk 1 above)
Comment (ian-r-rose) [inline doc/modules/changes.h]: Righto.

On Wed, May 27, 2015 at 7:42 PM, Wolfgang Bangerth <notifications@github.com

File: source/postprocess/boundary_densities.cc [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                          local_bottom_area += fe_face_values.JxW(q);
+                        }
+                  }
+              }
+
+      const double bottom_area = Utilities::MPI::sum( local_bottom_area, this->get_mpi_communicator() );
+      const double top_area = Utilities::MPI::sum( local_top_area, this->get_mpi_communicator() );
+
+      top_density = Utilities::MPI::sum( local_top_density, this->get_mpi_communicator() ) / top_area;
+      bottom_density = Utilities::MPI::sum( local_bottom_density, this->get_mpi_communicator() ) / bottom_area;
```
Comment (bangerth) [inline source/postprocess/boundary_densities.cc]: This is not efficient -- you are communicating four times here. Please create an array of four doubles, and do the sum only once via the function that does the sum on an array.

File: source/postprocess/boundary_densities.cc [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+
+#include <aspect/postprocess/boundary_densities.h>
+#include <aspect/simulator_access.h>
+
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/fe/fe_values.h>
+
+#include <boost/archive/text_oarchive.hpp>
+#include <boost/archive/text_iarchive.hpp>
```
Comment (bangerth) [inline source/postprocess/boundary_densities.cc]: I don't think you actually needs these two include files.

File: source/postprocess/boundary_pressures.cc [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+
+#include <aspect/postprocess/boundary_pressures.h>
+#include <aspect/simulator_access.h>
+
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/fe/fe_values.h>
+
+#include <boost/archive/text_oarchive.hpp>
+#include <boost/archive/text_iarchive.hpp>
```
Comment (bangerth) [inline source/postprocess/boundary_pressures.cc]: all the same comments for this file as for the previous one.

File: source/postprocess/boundary_densities.cc [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+                      for ( unsigned int q = 0; q < fe_face_values.n_quadrature_points; ++q)
+                        {
+                          local_bottom_density += out.densities[q] * fe_face_values.JxW(q);
+                          local_bottom_area += fe_face_values.JxW(q);
+                        }
+                  }
+              }
+
+      std::vector<double> values(4);  //vector for packing local values before mpi summing them
+      values[0] = local_bottom_area;
```
Comment (bangerth) [inline source/postprocess/boundary_densities.cc]: Still efficient. Do it as
```
  double values[4] = { local_bottom_area, local_top_area, ... };
```
This requires no dynamic memory allocation.

**Comment by tjhei** [comment]
Looks good. :+1:

**Comment by ian-r-rose** [comment]
Okay, hopefully this should be good to go

**Pull request merged**